// @sap/cds-lsp/lib/impl/prettyprint/formatOptions/SampleCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cdsFormatOptionsSampleCode = `using {
  Person    as User,
  UserIdent as UserID
} from './persons';

@requires             : 'authenticated-user'  @insertonly
@PropertyRestrictions : true                  @Searchable : true
entity Task : managed {
  key id       : /* UUID */ String(20) as taskId;
      assignee :            Association to one User @cds.on.insert /*1*/ :       #user
      priority :            Integer    as prio;
      dueOn    :            DateTime                @cds.on.insert       : /*2*/ #now;
}

context Lists {
  entity TaskList    as
      select from Task as task
      join ToDo as todo
        on  todo.taskId =      't' + task.id
        and task.dueOn  is not null
      mixin {
        taskType        : Association to one TaskType

                            on taskType.available = true;
        current         : Boolean = true;
        rejectionStatus : String  = 'none';
      }
      into {
        task.id : String,
        task.dueOn    as dueOn,
        task.assignee as assignee
      }
    union
      select from GenericTasks
      mixin {}
      into {
        'none' as id
      };

  entity HiddenTasks as
    select from Hidden {


      // only expose ID
      id
    }
}

type Number : Integer;
type Text : String;
`;
