// @sap/cds-lsp/lib/impl/utils/trace/Ansi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Ansi {
}
exports.Ansi = Ansi;
Ansi.Foreground = {
    BACK: '\x1b[30m',
    RED: '\x1b[31m',
    GREEN: '\x1b[32m',
    BROWN: '\x1b[33m',
    BLUE: '\x1b[34m',
    PURPLE: '\x1b[35m',
    DARKCYAN: '\x1b[36m',
    LIGHTGREY: '\x1b[37m',
    DARKGREY: '\x1b[1;30m',
    LIGHTRED: '\x1b[1;31m',
    LIGHTGREEN: '\x1b[1;32m',
    YELLOW: '\x1b[1;33m',
    LIGHTBLUE: '\x1b[1;34m',
    MAGENTA: '\x1b[1;35m',
    CYAN: '\x1b[1;36m',
    WHITE: '\x1b[1;37m'
};
Ansi.Background = {
    BLACK: '\x1b[40m',
    RED: '\x1b[41m',
    GREEN: '\x1b[42m',
    BROWN: '\x1b[43m',
    BLUE: '\x1b[44m',
    PURPLE: '\x1b[45m',
    DARKCYAN: '\x1b[46m',
    LIGHTGREY: '\x1b[47m',
};
Ansi.ITALIC = '\x1b[3m';
Ansi.RESET = '\x1b[0m' + Ansi.Background.BLACK;
Ansi.BLOCK = Ansi.ITALIC;
Ansi.REQUEST_FROM_SERVER = Ansi.Foreground.PURPLE;
Ansi.RESPONSE_TO_CLIENT = Ansi.Foreground.LIGHTGREEN;
Ansi.CATEGORY = Ansi.RESET + Ansi.Foreground.LIGHTGREY;
Ansi.TOKEN = '\x1b[0m\x1b[35m';
Ansi.COMMENT_TOKEN = Ansi.RESET + Ansi.Foreground.MAGENTA;
Ansi.RULE = Ansi.RESET + Ansi.Foreground.GREEN;
Ansi.ERROR = Ansi.RESET + Ansi.Background.RED + Ansi.Foreground.YELLOW;
Ansi.WARN = Ansi.RESET + Ansi.Foreground.YELLOW;
Ansi.INFO = Ansi.RESET + Ansi.Foreground.WHITE;
Ansi.SUCCESS = Ansi.RESET + Ansi.Background.GREEN;
Ansi.KEYWORD = '\x1b[0m';
Ansi.KNOWN_DEFINITION = '\x1b[0m\x1b[32m';
Ansi.KNOWN_IDENTIFIER = '\x1b[0m\x1b[1;34m';
Ansi.WEAK_IDENTIFIER = '\x1b[0m\x1b[34m';
Ansi.UNKNOWN_IDENTIFIER = '\x1b[0m\x1b[1;31m';
Ansi.KIND = '\x1b[0m\x1b[37m';
