// @sap/cds-lsp/lib/impl/utils/trace/ContinuationLocalStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const async_hooks = require("async_hooks");
class ClsFrame {
    constructor(asyncId) {
        this.asyncId = asyncId;
    }
}
exports.ClsFrame = ClsFrame;
class ContinuationLocalStorage {
    constructor(clsFactory) {
        this.frames = new Map();
        this.clsFactory = clsFactory;
        if (!this.hook) {
            this.hook = async_hooks.createHook({
                init: (asyncId, type, triggerAsyncId, resource) => {
                    const parent = this.getFrame(triggerAsyncId);
                    const newChild = this.getFrame(asyncId, parent);
                    this.frames.set(asyncId, newChild);
                },
                destroy: (asyncId) => {
                    this.frames.delete(asyncId);
                }
            });
            this.hook.enable();
        }
    }
    getFrame(asyncId, template) {
        let frame = this.frames.get(asyncId);
        if (!frame) {
            frame = this.clsFactory.createFrame(asyncId, template);
            this.frames.set(asyncId, frame);
        }
        return frame;
    }
    currentFrame() {
        return this.getFrame(async_hooks.executionAsyncId());
    }
}
exports.ContinuationLocalStorage = ContinuationLocalStorage;
