// @sap/cds-lsp/lib/impl/prettyprint/util/Ansi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Ansi_Foreground;
(function (Ansi_Foreground) {
    Ansi_Foreground["BLACK"] = "\u001B[30m";
    Ansi_Foreground["RED"] = "\u001B[31m";
    Ansi_Foreground["GREEN"] = "\u001B[32m";
    Ansi_Foreground["BROWN"] = "\u001B[33m";
    Ansi_Foreground["BLUE"] = "\u001B[34m";
    Ansi_Foreground["PURPLE"] = "\u001B[35m";
    Ansi_Foreground["DARKCYAN"] = "\u001B[36m";
    Ansi_Foreground["LIGHTGREY"] = "\u001B[37m";
    Ansi_Foreground["DARKGREY"] = "\u001B[1;30m";
    Ansi_Foreground["LIGHTRED"] = "\u001B[1;31m";
    Ansi_Foreground["LIGHTGREEN"] = "\u001B[1;32m";
    Ansi_Foreground["YELLOW"] = "\u001B[1;33m";
    Ansi_Foreground["LIGHTBLUE"] = "\u001B[1;34m";
    Ansi_Foreground["MAGENTA"] = "\u001B[1;35m";
    Ansi_Foreground["CYAN"] = "\u001B[1;36m";
    Ansi_Foreground["WHITE"] = "\u001B[1;37m";
})(Ansi_Foreground = exports.Ansi_Foreground || (exports.Ansi_Foreground = {}));
var Ansi_Background;
(function (Ansi_Background) {
    Ansi_Background["BLACK"] = "\u001B[40m";
    Ansi_Background["RED"] = "\u001B[41m";
    Ansi_Background["GREEN"] = "\u001B[42m";
    Ansi_Background["BROWN"] = "\u001B[43m";
    Ansi_Background["BLUE"] = "\u001B[44m";
    Ansi_Background["PURPLE"] = "\u001B[45m";
    Ansi_Background["DARKCYAN"] = "\u001B[46m";
    Ansi_Background["LIGHTGREY"] = "\u001B[47m";
    Ansi_Background["DARKGREY"] = "\u001B[1;40m";
    Ansi_Background["LIGHTRED"] = "\u001B[1;41m";
    Ansi_Background["LIGHTGREEN"] = "\u001B[1;42m";
    Ansi_Background["YELLOW"] = "\u001B[1;43m";
    Ansi_Background["LIGHTBLUE"] = "\u001B[1;44m";
    Ansi_Background["MAGENTA"] = "\u001B[1;45m";
    Ansi_Background["CYAN"] = "\u001B[1;46m";
    Ansi_Background["WHITE"] = "\u001B[1;47m";
})(Ansi_Background = exports.Ansi_Background || (exports.Ansi_Background = {}));
class Ansi {
}
exports.Ansi = Ansi;
Ansi.ITALIC = '\x1b[3m';
Ansi.RESET = '\x1b[0m';
Ansi.BLOCK = Ansi.ITALIC;
Ansi.CATEGORY = Ansi.RESET + Ansi_Foreground.LIGHTGREY;
Ansi.TOKEN = Ansi.RESET + Ansi_Foreground.PURPLE;
Ansi.ERROR_TOKEN = Ansi.RESET + Ansi_Foreground.RED;
Ansi.COMMENT_TOKEN = Ansi.RESET + Ansi_Foreground.YELLOW;
Ansi.RULE = Ansi.RESET + Ansi_Foreground.GREEN;
Ansi.ERROR = Ansi.RESET + Ansi_Background.LIGHTRED + Ansi_Foreground.YELLOW;
Ansi.WARN = Ansi.RESET + Ansi_Background.YELLOW;
Ansi.INFO = Ansi.RESET + Ansi_Background.LIGHTGREEN;
