// @sap/cds-lsp/lib/impl/prettyprint/util/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Ansi_1 = require("./Ansi");
class Logger {
    isTracing(pp) {
        return false;
    }
    trace(...args) { }
    logError(e) { }
    logWarn(e) { }
    logInfo(s) { }
    logVerbose(s) { }
    logDebug(s) { }
}
exports.Logger = Logger;
var Color;
(function (Color) {
    Color[Color["BLOCK"] = 0] = "BLOCK";
    Color[Color["CATEGORY"] = 1] = "CATEGORY";
    Color[Color["TOKEN"] = 2] = "TOKEN";
    Color[Color["ERROR_TOKEN"] = 3] = "ERROR_TOKEN";
    Color[Color["COMMENT_TOKEN"] = 4] = "COMMENT_TOKEN";
    Color[Color["RULE"] = 5] = "RULE";
    Color[Color["ERROR"] = 6] = "ERROR";
    Color[Color["WARN"] = 7] = "WARN";
    Color[Color["INFO"] = 8] = "INFO";
    Color[Color["RESET"] = 9] = "RESET";
})(Color = exports.Color || (exports.Color = {}));
class AnsiColorProvider {
    get(c) {
        switch (c) {
            case Color.WARN: return Ansi_1.Ansi.WARN;
            case Color.TOKEN: return Ansi_1.Ansi.TOKEN;
            case Color.INFO: return Ansi_1.Ansi.INFO;
            case Color.ERROR_TOKEN: return Ansi_1.Ansi.ERROR_TOKEN;
            case Color.ERROR: return Ansi_1.Ansi.ERROR;
            case Color.COMMENT_TOKEN: return Ansi_1.Ansi.COMMENT_TOKEN;
            case Color.BLOCK: return Ansi_1.Ansi.BLOCK;
            case Color.RULE: return Ansi_1.Ansi.RULE;
            case Color.CATEGORY: return Ansi_1.Ansi.CATEGORY;
            case Color.RESET: return Ansi_1.Ansi.RESET;
            default: return '';
        }
    }
}
exports.AnsiColorProvider = AnsiColorProvider;
class NoColorProvider {
    get(c) {
        return '';
    }
}
exports.NoColorProvider = NoColorProvider;
function redirectLogger(tracer, logger) {
    logger.logError = (s) => tracer.error(s);
    logger.logWarn = (s) => tracer.warn(s);
    logger.logInfo = (s) => tracer.info(s);
    logger.logVerbose = (s) => tracer.verbose(s);
    logger.logDebug = (s) => tracer.debug(s);
    logger.trace = (...args) => tracer.verbose(args.join(' '));
    return logger;
}
exports.redirectLogger = redirectLogger;
