// @sap/cds-lsp/lib/impl/prettyprint/FacsimileAlignment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Alignment_1 = require("./Alignment");
class FacsimileAlignment extends Alignment_1.Alignment {
    constructor(context, firstColumn) {
        super(context, '_fa_' + context.nextFreeAlignmentId);
        this.firstColumn = firstColumn;
        this.relativeColumns = [0];
    }
    toString() {
        return `FA(${this.id})`;
    }
    registerFirstHardenedLine(lineNumber) {
        if (this.firstHardenedLineNumber === undefined || lineNumber < this.firstHardenedLineNumber) {
            this.firstHardenedLineNumber = lineNumber;
        }
    }
    registerFirstHardenedColumn(lineNumber, column) {
        if (lineNumber === this.firstHardenedLineNumber) {
            this.firstHardenedColumn = column;
        }
    }
    setIndentation(lineOffset, column) {
        this.relativeColumns[lineOffset] = column - this.firstColumn;
    }
    getColumn(hardenedLineNumber) {
        return this.firstHardenedColumn + this.relativeColumns[hardenedLineNumber - this.firstHardenedLineNumber];
    }
}
exports.FacsimileAlignment = FacsimileAlignment;
