// @sap/cds-lsp/lib/impl/services/refactor/RefactoringService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../utils");
const trace_1 = require("../../utils/trace");
class RefactoringService {
    constructor() {
        this.tracer = trace_1.currentTrace().tracer(trace_1.LspTraceComponent.REFACTOR);
    }
    onRenameRequest(params) {
        return null;
    }
}
__decorate([
    trace_1.RequestMethod({
        paramsRenderer: utils_1.renderAsIs,
        resultRenderer: utils_1.renderAsIs,
        onErrorReturn: utils_1.empty()
    })
], RefactoringService.prototype, "onRenameRequest", null);
exports.RefactoringService = RefactoringService;
